### PR TITLE
Reduce memory usage by applying query check sooner

### PR DESCRIPTION
**Pull request by wu-hui**
see: https://github.com/firebase/firebase-android-sdk/pull/4591

### CODE REVIEW

File: Firestore/core/src/local/leveldb_remote_document_cache.cc [hunk 1]
```diff
@@ -171,13 +173,18 @@ MutableDocumentMap LevelDbRemoteDocumentCache::GetAll(
...
 MutableDocumentMap LevelDbRemoteDocumentCache::GetAllExisting(
-    DocumentVersionMap&& remote_map) const {
+    DocumentVersionMap&& remote_map,
+    const core::Query& query,
+    const model::OverlayByDocumentKeyMap& mutated_docs) const {
   BackgroundQueue tasks(executor_.get());
   AsyncResults<std::pair<DocumentKey, MutableDocument>> results;
   for (const auto& key_version : remote_map) {
-    tasks.Execute([this, &results, key_version] {
+    tasks.Execute([this, &results, key_version, query, mutated_docs] {
```
Comment (cherylEnkidu) [inline Firestore/core/src/local/leveldb_remote_document_cache.cc:182]: can we pass `query` and `mutated_docs` by reference?

**Review by cherylEnkidu — CHANGES_REQUESTED** [review]
Thank you for the improvement!

File: Firestore/core/src/local/leveldb_remote_document_cache.cc [hunk 2]
```diff
@@ -171,13 +173,18 @@ MutableDocumentMap LevelDbRemoteDocumentCache::GetAll(
...
 MutableDocumentMap LevelDbRemoteDocumentCache::GetAllExisting(
-    DocumentVersionMap&& remote_map) const {
+    DocumentVersionMap&& remote_map,
+    const core::Query& query,
+    const model::OverlayByDocumentKeyMap& mutated_docs) const {
   BackgroundQueue tasks(executor_.get());
   AsyncResults<std::pair<DocumentKey, MutableDocument>> results;
   for (const auto& key_version : remote_map) {
-    tasks.Execute([this, &results, key_version] {
+    tasks.Execute([this, &results, key_version, &query, &mutated_docs] {
```
Comment (cherylEnkidu) [inline Firestore/core/src/local/leveldb_remote_document_cache.cc:182]: Sorry I didn't find out earlier, could you please also make `key_version` pass by reference?

**Review by cherylEnkidu — APPROVED** [review]

**Comment by cherylEnkidu** [comment]
Looks like `sanitizers` complains after merge the master branch

File: Firestore/core/src/local/leveldb_remote_document_cache.cc (hunk 2 above)
Comment (wu-hui) [reply Firestore/core/src/local/leveldb_remote_document_cache.cc]: I do not think query can be captured by reference here, somehow it is deleted before its usage here. This is what caused sanitizer failure.

Comment (cherylEnkidu) [reply Firestore/core/src/local/leveldb_remote_document_cache.cc]: I don't understand why `query` would be deleted :(

**Pull request merged**